dpsprep.1
.\" generated with Ronn-NG/v0.10.1
.\" http://github.com/apjanke/ronn-ng/tree/0.10.1
.TH "DPS" "1" "March 2026" ""
.SH "NAME"
\fBdps\fR \- a DjVu to PDF converter
.SH "SYNOPSIS"
\fBdpsprep\fR \fIoptions\fR src [dest]
.SH "DESCRIPTION"
This tool, initially made specifically for use with Sony's Digital Paper System (DPS), is now a general\-purpose DjVu to PDF converter with a focus on small output size and the ability to preserve document outlines (e\.g\. TOC) and text layers (e\.g\. OCR)\.
.SH "OPTIONS"
.IP "\(bu" 4
\fB\-q\fR, \fB\-\-quality\fR: Quality of images in output\. Used only for JPEG compression, i\.e\. RGB and Grayscale images\. Passed directly to Pillow and to OCRmyPDF's optimizer\.
.IP "\(bu" 4
\fB\-v\fR, \fB\-\-verbose\fR: Display debug messages\.
.IP "\(bu" 4
\fB\-o\fR, \fB\-\-overwrite\fR: Overwrite destination file\.
.IP "\(bu" 4
\fB\-w\fR, \fB\-\-preserve\-working\fR: Preserve the working directory after script termination\.
.IP "\(bu" 4
\fB\-d\fR, \fB\-\-delete\-working\fR: Delete any existing files in the working directory prior to writing to it\.
.IP "\(bu" 4
\fB\-t\fR, \fB\-\-no\-text\fR: Disable the generation of text layers\. Implied by \-\-ocr\.
.IP "\(bu" 4
\fB\-p\fR, \fB\-\-pool\-size\fR \fIint\fR: Size of MultiProcessing pool for handling page\-by\-page operations\.
.IP "\(bu" 4
\fB\-m\fR, \fB\-\-mode\fR \fIinfer|bitonal|grayscale|rgb\fR: Override the image modes encoded in the DjVu file for individual pages\. It sometimes makes sense to force bitonal images since they compress well\.
.IP "\(bu" 4
\fB\-\-dpi\fR \fIint\fR: Override DPI values encoded in the DjVu file for individual pages\.
.IP "\(bu" 4
\fB\-\-ocr\fR \fIJSON\fR: Perform OCR via OCRmyPDF rather than trying to convert the text layer\. If this parameter has a value, it should be a JSON dictionary of options to be passed to OCRmyPDF\.
.IP "\(bu" 4
\fB\-O1\fR: Use the lossless PDF image optimization from OCRmyPDF (without performing OCR)\.
.IP "\(bu" 4
\fB\-O2\fR: Use the PDF image optimization from OCRmyPDF\.
.IP "\(bu" 4
\fB\-O3\fR: Use the aggressive lossy PDF image optimization from OCRmyPDF\.
.IP "\(bu" 4
\fB\-\-help\fR: Show help message and exit\.
.IP "\(bu" 4
\fB\-\-version\fR: Show the version and exit\.
.IP "" 0
.SH "EXAMPLES"
Produce \fBfile\.pdf\fR in the current directory:
.IP "" 4
.nf
dpsprep /wherever/file\.djvu
.fi
.IP "" 0
.P
Produce \fBoutput\.pdf\fR with reduced image quality and aggressive PDF image optimizations:
.IP "" 4
.nf
dpsprep \-\-\-quality=30 \-O3 input\.djvu output\.pdf
.fi
.IP "" 0
.P
Produce an output file using a large pool of workers:
.IP "" 4
.nf
dpsprep \-\-pool=16 input\.djvu
.fi
.IP "" 0
.P
Force bitonal images:
.IP "" 4
.nf
dpsprep \-\-mode bitonal input\.djvu
.fi
.IP "" 0
.P
Produce an output file by disregarding the text layer and running OCRmyPDF instead:
.IP "" 4
.nf
dpsprep \-\-ocr '{"language": ["rus", "eng"]}' input\.djvu
.fi
.IP "" 0
.P
Or simply disregard the text layer without OCR:
.IP "" 4
.nf
dpsprep \-\-no\-text input\.djvu
.fi
.IP "" 0
.SH "NOTE REGARDING COMPRESSION"
We perform compression in two stages:
.IP "\(bu" 4
The first one is the default compression provided by Pillow\. For bitonal images, the PDF generation code says that, if \fBlibtiff\fR is available, \fBgroup4\fR compression is used\.
.IP "\(bu" 4
If OCRmyPDF is installed, its PDF optimization can be used via the flags \fB\-O1\fR to \fB\-O3\fR (this involves no OCR)\. This allows us to use advanced techniques, including JBIG2 compression via \fBjbig2enc\fR\.
.IP "" 0
.P
If manually running OCRmyPDF, note that the optimization command suggested in the documentation (setting \fB\-\-tesseract\-timeout\fR to \fB0\fR) may ruin existing text layers\. To perform only PDF optimization you can use the following undocumented tool instead:
.IP "" 4
.nf
python \-m ocrmypdf\.optimize <input_file> <level> <output_file>
.fi
.IP "" 0



CHANGELOG.md
## v2.5.1 (2026-03-14)

* Allow manually configuring PDF page resolution (DPI)

## v2.5.0 (2026-03-13)

* Account for DjVu file resolution
* Simplify image diffing and regenerate better-quality fixtures

## v2.4.2 (2026-02-24)

* Fix issue where only the main process has its logger configured

## v2.4.1 (2026-02-24)

* Fix compatibility issues with the new OCRmyPDF API
* Remove support for Python 3.10

## v2.4.0 (2026-02-24)

* Migrate to `uv` from `pyenv` + `poetry`
* Update dependencies

## v2.3.1 (2025-10-28)

* Fix mixed-up email format

## v2.3.0 (2025-10-28)

* Remove support for Python 3.9
* Migrate to standardized `pyproject.toml`
* Update dependencies

## v2.2.15 (2025-07-02)

* Add support for installation via `pipx`

## v2.2.14 (2025-05-27)

* Improve installation notes
* Bump djvulibre-python version

## v2.2.13 (2025-02-12)

* Fail-safe quality settings for non-JPEG images

## v2.2.12 (2025-01-27)

* Update pytest_image_diff and fix newly broken tests

## v2.2.11 (2025-01-26)

* Update dependencies

## v2.2.10 (2024-10-25)

* Improve interface with OCRmyPDF
* Fix CI build

## v2.2.9 (2024-10-25)

* Improve type hints
* Update dependencies

## v2.2.8 (2024-10-18)

* Support single characters in the text layer

## v2.2.7 (2024-08-27)

* Improve tab and newline handling

## v2.2.6 (2024-08-05)

* Fix accidental whitespace removal from text blocks

## v2.2.5 (2024-07-20)

* Re-add ability to force the image mode (RGB/Grayscale/Monochrome)

## v2.2.4 (2024-02-24)

* Update dependencies

## v2.2.3 (2023-12-09)

* Fix CI build
* Ignore invalid UTF-8 sequences
* Ignore unrecognized page titles in the outline (#23)

## v2.2.2 (2023-10-29)

* Update dependencies

## v2.2.1 (2023-11-06)

* Handle invalid PDF pages
* Fix exception in text layer processing (#20)

## v2.2.0 (2023-10-28)

* Add options for disabling the text layer and for directly running OCR

## v2.1.5 (2023-10-27)

* Fix inverted colors in images (#16)

## v2.1.4 (2023-10-06)

* Fix typo in logging code

## v2.1.3 (2023-10-06)

* Improve logging

## v2.1.2 (2023-10-02)

* Accidental version bump

## v2.1.1 (2023-10-02)

* Remove debug code

## v2.1.0 (2023-10-02)

* Add support for OCRmyPDF

## v2.0.2 (2023-08-03)

* Update some other dependencies
* Replace `python-djvulibre` with `djvulibre-python`

## v2.0.1 (2023-06-22)

* Minor improvements in packaging

## v2.0.0 (2023-05-04)

* Fully rewrite


dpsprep/dpsprep.py
import json
import multiprocessing.pool
import shutil
from time import time

import click
import djvu.decode
import loguru
import pdfrw

from .images import ImageMode, failsafe_save_djvu_page, process_djvu_page
from .logging import configure_loguru, human_readable_size
from .ocrmypdf import optimize_pdf, perform_ocr
from .outline import OutlineTransformVisitor
from .pdf import combine_pdfs_on_fs_with_text, combine_pdfs_on_fs_without_text, is_valid_pdf
from .text import djvu_pages_to_text_fpdf
from .workdir import WorkingDirectory


def process_page_bg(workdir: WorkingDirectory, mode: ImageMode, quality: int | None, dpi: int | None, i: int, *, verbose: bool) -> None:  # noqa: PLR0913
    configure_loguru(verbose=verbose)
    page_number = i + 1

    if workdir.get_page_pdf_path(i).exists():
        if is_valid_pdf(workdir.get_page_pdf_path(i)):
            loguru.logger.debug(f'Image data from page {page_number} already processed.')
            return
        loguru.logger.debug(f'Invalid page generated for {page_number}, regenerating.')
    else:
        loguru.logger.debug(f'Processing image data from page {page_number}.')

    start_time = time()
    document = djvu.decode.Context().new_document(
        djvu.decode.FileURI(workdir.src),
    )
    document.decoding_job.wait()

    page_bg = process_djvu_page(document.pages[i], mode, i)

    failsafe_save_djvu_page(
        page_bg,
        workdir.get_page_pdf_path(i),
        quality,
        dpi,
        page_number,
    )

    pdf_size = workdir.get_page_pdf_path(i).stat().st_size
    loguru.logger.debug(f'Image data with size {human_readable_size(pdf_size)} from page {page_number} processed in {time() - start_time:.2f}s and written to working directory.')


def process_text(workdir: WorkingDirectory, dpi: int | None, *, verbose: bool) -> None:
    configure_loguru(verbose=verbose)

    if workdir.text_layer_pdf_path.exists():
        loguru.logger.info('Text data already processed.')
        return

    loguru.logger.debug('Processing text data.')

    start_time = time()
    document = djvu.decode.Context().new_document(
        djvu.decode.FileURI(workdir.src),
    )
    document.decoding_job.wait()

    fpdf = djvu_pages_to_text_fpdf(document.pages, dpi)
    fpdf.output(str(workdir.text_layer_pdf_path))

    pdf_size = workdir.text_layer_pdf_path.stat().st_size
    loguru.logger.info(f'Text data with size {human_readable_size(pdf_size) } processed in {time() - start_time:.2f}s and written to working directory')


@click.option('-d', '--delete-working', is_flag=True, help='Delete any existing files in the working directory prior to writing to it.')
@click.option('-w', '--preserve-working', is_flag=True, help='Preserve the working directory after script termination.')
@click.option('-o', '--overwrite', is_flag=True, help='Overwrite destination file.')
@click.option('-v', '--verbose', is_flag=True, help='Display debug messages.')
@click.option('-t', '--no-text', is_flag=True, help='Disable the generation of text layers. Implied by --ocr.')
@click.option('-O1', 'optlevel', flag_value=1, help='Use the lossless PDF image optimization from OCRmyPDF (without performing OCR).')
@click.option('-O2', 'optlevel', flag_value=2, help='Use the PDF image optimization from OCRmyPDF.')
@click.option('-O3', 'optlevel', flag_value=3, help='Use the aggressive lossy PDF image optimization from OCRmyPDF.')
@click.option('-p', '--pool-size', type=click.IntRange(min=0), default=4, help='Size of MultiProcessing pool for handling page-by-page operations.')
@click.option('-q', '--quality', type=click.IntRange(min=0, max=100), help="Quality of images in output. Used only for JPEG compression, i.e. RGB and Grayscale images. Passed directly to Pillow and to OCRmyPDF's optimizer.")
@click.option('-m', '--mode', type=click.Choice(['infer', 'bitonal', 'grayscale', 'rgb']), default='infer', help='Override the image modes encoded in the DjVu file for individual pages. It sometimes makes sense to force bitonal images since they compress well.')
@click.option('--dpi', type=click.IntRange(min=1), help='Override DPI values encoded in the DjVu file for individual pages.')
@click.option('--ocr', type=str, is_flag=False, flag_value='{}', help='Perform OCR via OCRmyPDF rather than trying to convert the text layer. If this parameter has a value, it should be a JSON dictionary of options to be passed to OCRmyPDF.')
@click.version_option()
@click.argument('dest', type=click.Path(exists=False, resolve_path=True), required=False)
@click.argument('src', type=click.Path(exists=True, resolve_path=True), required=True)
@click.command()
def dpsprep(  # noqa: C901, PLR0912, PLR0913, PLR0915
    src: str,
    dest: str | None,
    quality: int | None,
    dpi: int | None,
    pool_size: int,
    mode: ImageMode,
    optlevel: int | None,
    ocr: str | None,
    *,
    verbose: bool,
    overwrite: bool,
    delete_working: bool,
    preserve_working: bool,
    no_text: bool,
) -> None:
    configure_loguru(verbose=verbose)
    workdir = WorkingDirectory(src, dest)

    if ocr is None:
        ocr_options = None
    else:
        try:
            ocr_options = json.loads(ocr)
        except ValueError as err:
            msg = f'The OCR options {ocr!r} are not valid JSON.'
            raise SystemExit(msg) from err
        else:
            if not isinstance(ocr_options, dict):
                msg = f'The OCR options {ocr!r} are not a JSON dictionary.'
                raise SystemExit(msg)

        no_text = True

    if not overwrite and workdir.dest.exists():
        msg = f'File {workdir.dest} already exists.'
        raise SystemExit(msg)

    start_time = time()

    if workdir.workdir.exists():
        if delete_working:
            loguru.logger.debug(f'Removing existing working directory {workdir.workdir}.')
            workdir.destroy()
            loguru.logger.info(f'Removed existing working directory {workdir.workdir}.')
        else:
            loguru.logger.info(f'Reusing working directory {workdir.workdir}.')
    else:
        loguru.logger.info(f'Working directory {workdir.workdir} has been created.')

    workdir.create_if_necessary()

    document = djvu.decode.Context().new_document(
        djvu.decode.FileURI(workdir.src),
    )
    document.decoding_job.wait()

    djvu_size = workdir.src.stat().st_size
    loguru.logger.info(f'Processing {workdir.src} with {len(document.pages)} pages and size {human_readable_size(djvu_size)} using {pool_size} workers.')

    pool = multiprocessing.Pool(processes=pool_size)
    tasks = list[multiprocessing.pool.AsyncResult]()

    if not no_text:
        tasks.append(pool.apply_async(func=process_text, args=[workdir], kwds={'verbose': verbose}))

    for i in range(len(document.pages)):
        # Cannot pass the page object itself because it does not support serialization for IPC
        tasks.append(pool.apply_async(func=process_page_bg, args=[workdir, mode, quality, dpi, i], kwds={'verbose': verbose}))

    pool.close()
    pool_is_working = True

    while pool_is_working:
        pool_is_working = False

        for task in tasks:
            try:
                task.get(timeout=25)
            except multiprocessing.TimeoutError:
                pool_is_working = True

    pool.join()
    loguru.logger.info('Processed all pages.')

    outline = pdfrw.IndirectPdfDict()

    if len(document.outline.sexpr) > 0:
        loguru.logger.info('Processing metadata.')
        outline = OutlineTransformVisitor().visit(document.outline.sexpr)
        loguru.logger.info('Metadata processed.')
    else:
        loguru.logger.info('No metadata to process.')

    loguru.logger.info('Combining everything.')

    if no_text:
        combine_pdfs_on_fs_without_text(workdir, outline, len(document.pages))

        ocr_success = False

        if ocr_options:
            loguru.logger.info('Performing OCR.')
            ocr_success = perform_ocr(workdir, ocr_options)
        else:
            loguru.logger.info('Skipping the text layer.')

        if not ocr_success:
            shutil.copy(workdir.combined_pdf_without_text_path, workdir.combined_pdf_path)
    else:
        combine_pdfs_on_fs_with_text(workdir, outline)

    combined_size = workdir.combined_pdf_path.stat().st_size
    loguru.logger.info(f'Produced a combined output file with size {human_readable_size(combined_size)} in {time() - start_time:.2f}s. This is {round(100 * combined_size / djvu_size, 2)}% of the DjVu source file.')

    opt_success = False

    if optlevel is not None:
        loguru.logger.info(f'Performing level {optlevel} optimization.')
        opt_success = optimize_pdf(workdir, optlevel, quality, pool_size)

    if opt_success:
        opt_size = workdir.optimized_pdf_path.stat().st_size

        loguru.logger.info(f'The optimized file has size {human_readable_size(opt_size)}, which is {round(100 * opt_size / combined_size, 2)}% of the raw combined file and {round(100 * opt_size / djvu_size, 2)}% of the DjVu source file.')

        if opt_size < combined_size:
            loguru.logger.info('Using the optimized file.')
            shutil.copy(workdir.optimized_pdf_path, workdir.dest)
        else:
            loguru.logger.info('Using the raw combined file.')
            shutil.copy(workdir.combined_pdf_path, workdir.dest)
    else:
        shutil.copy(workdir.combined_pdf_path, workdir.dest)

    if preserve_working:
        loguru.logger.info(f'Working directory {workdir.workdir} will be preserved.')
    else:
        loguru.logger.info(f'Deleting the working directory {workdir.workdir}.')
        workdir.destroy()
